### PR TITLE
Fixes to properly free allocated memory at shutdown

### PR DESCRIPTION
**Pull request by ntfshard**

### CODE REVIEW

File: alica_engine/src/engine/PlanBase.cpp
Comment (Dale-Koenig) [inline alica_engine/src/engine/PlanBase.cpp]: Is this something that can be a unique ptr?

Comment (ntfshard) [reply alica_engine/src/engine/PlanBase.cpp]: I guess so, but out of scope 
Hope it won't break something in destruction order

File: alica_engine/src/engine/PlanBase.cpp
Comment (Dale-Koenig) [inline alica_engine/src/engine/PlanBase.cpp:343]: Nit: not needed

File: supplementary/alica_ros1/alica_tests/src/test/test_assignment.cpp
Comment (Dale-Koenig) [inline supplementary/alica_ros1/alica_tests/src/test/test_assignment.cpp:98]: Destruction doesn't terminate?

**Review by Dale-Koenig — COMMENTED** [review]
CI is failing.  Are you still updating this PR?

File: supplementary/alica_ros1/alica_tests/src/test/test_assignment.cpp
Comment (ntfshard) [reply supplementary/alica_ros1/alica_tests/src/test/test_assignment.cpp:98]: Seems no. AFAIR adding this line changes the valgrind's results.

File: alica_engine/src/engine/PlanBase.cpp
Comment (ntfshard) [reply alica_engine/src/engine/PlanBase.cpp:343]: Just keep the previous behaviour of the code. And protection from error in case of calling `stop` twice in a row.

**Comment by ntfshard** [comment]
> CI is failing. Are you still updating this PR?

Yes, faced with some interesting problems

**Review by Dale-Koenig — DISMISSED** [review]

File: alica_engine/src/engine/AlicaContext.cpp
Comment (ntfshard) [inline alica_engine/src/engine/AlicaContext.cpp]: I guess we should call `AlicaContext::terminate()` if it wasn't called outside

File: alica_dummy_proxy/src/communication/AlicaDummyCommunication.cpp
Comment (ntfshard) [inline alica_dummy_proxy/src/communication/AlicaDummyCommunication.cpp:221]: I guess if we use this class only for testign, I guess we should handle `tick()` call, it works like epoch sync here, AFAIU

File: alica_engine/include/engine/AlicaContext.h
Comment (ntfshard) [inline alica_engine/include/engine/AlicaContext.h:428]: It feels like a hack, I hope we can find way to get rid of it

File: alica_engine/src/engine/PlanBase.cpp
Comment (ntfshard) [inline alica_engine/src/engine/PlanBase.cpp:337]: This place should be groomed

**Review by ntfshard — COMMENTED** [review]
yeah...

File: alica_engine/src/engine/modelmanagement/factories/VariableFactory.cpp
Comment (bjoernschroeder) [inline alica_engine/src/engine/modelmanagement/factories/VariableFactory.cpp:20]: Where is the memory deallocated? I could not find anything in Factory, ModelManager and PlanRepository.

Comment (ntfshard) [reply alica_engine/src/engine/modelmanagement/factories/VariableFactory.cpp:20]: it stored in unique_ptr by Factory::storeElement function.

https://github.com/rapyuta-robotics/alica/pull/388/files/b53ba625d316b59b38ab9cf2731b0d84629b0d53#diff-5193dd514b3d802792d71a4d2c982c6f1681f51345bc7f8f174d54ae44787557R125

Comment (bjoernschroeder) [reply alica_engine/src/engine/modelmanagement/factories/VariableFactory.cpp:20]: Just saw that you wrapped it in a unique_ptr in storeElement now, so memory should be deallocated when the modelManager is destroyed

File: supplementary/alica_ros1/alica_tests/src/test/test_assignment.cpp
Comment (ntfshard) [reply supplementary/alica_ros1/alica_tests/src/test/test_assignment.cpp:98]: no it will be called by dtor if it was initialized but not terminated

File: alica_engine/src/engine/AlicaContext.cpp
Comment (ntfshard) [reply alica_engine/src/engine/AlicaContext.cpp]: Changed

File: alica_engine/src/engine/PlanBase.cpp
Comment (triccyx) [inline alica_engine/src/engine/PlanBase.cpp:158]: `load` or `std::atomic<T>::operator T` should be the same. Am I missing something?

File: alica_engine/src/engine/PlanBase.cpp
Comment (triccyx) [inline alica_engine/src/engine/PlanBase.cpp:343]: I prefer `_mainThread=nullptr` because it is more clear. But it is just my preference. I think it is the same.

**Review by triccyx — COMMENTED** [review]
A couple of minor

File: alica_engine/src/engine/PlanBase.cpp
Comment (ntfshard) [reply alica_engine/src/engine/PlanBase.cpp:158]: `std::atomic<T>::operator T` [equivalent ](https://en.cppreference.com/w/cpp/atomic/atomic/operator_T)to `load()`.

Here we should help compiler with lamba return type deduction

File: alica_engine/src/engine/PlanBase.cpp
Comment (ntfshard) [reply alica_engine/src/engine/PlanBase.cpp:343]: AFAIS unique_ptr::operator= has special overload for nullptr. Such fancy

File: alica_engine/src/engine/PlanBase.cpp
Comment (ntfshard) [reply alica_engine/src/engine/PlanBase.cpp:158]: I guess we could call `load(memory_order::memory_order_relaxed)`  but not sure that we need it

File: alica_engine/src/engine/modelmanagement/factories/VariableFactory.cpp
Comment (xffox) [inline alica_engine/src/engine/modelmanagement/factories/VariableFactory.cpp:20]: Can you create `std::unique_ptr` already here, seems safer?

File: alica_engine/src/engine/RuntimePlanFactory.cpp
Comment (xffox) [inline alica_engine/src/engine/RuntimePlanFactory.cpp]: Nit: can use `wait()` with predicate.

File: alica_engine/src/engine/RuntimePlanFactory.cpp
Comment (xffox) [inline alica_engine/src/engine/RuntimePlanFactory.cpp:44]: Can probably unlock at this point.

File: alica_engine/src/engine/syncmodule/SyncModule.cpp
Comment (xffox) [inline alica_engine/src/engine/syncmodule/SyncModule.cpp:41]: Seems suspicious. I think if you need to lock in the destructor this object can be accessed while being destroyed, wouldn't this be forbidden?

Comment (ntfshard) [reply alica_engine/src/engine/syncmodule/SyncModule.cpp:41]: I made it for consistency, hope we don't have a situation when we are trying to destroy an object while other thread a inside the method

File: alica_engine/src/engine/modelmanagement/factories/VariableFactory.cpp
Comment (ntfshard) [reply alica_engine/src/engine/modelmanagement/factories/VariableFactory.cpp:20]: unique_ptr will be constructed inside `Factory::storeElement`

in other words, to change here to unique_ptr we should make a refactoring of all factories.

File: alica_engine/src/engine/RuntimePlanFactory.cpp
Comment (ntfshard) [reply alica_engine/src/engine/RuntimePlanFactory.cpp:44]: For what effect we should unlock at this point?
We need this sync because we starting communication before we complete whole initialization and we have cases when we have calls here before we receive _creator. If we somehow can get a recursive call here, I really want to know where

File: alica_engine/src/engine/PlanBase.cpp
Comment (triccyx) [reply alica_engine/src/engine/PlanBase.cpp:158]: Understood ok for .load for me

**Review by triccyx — APPROVED** [review]

File: alica_engine/src/engine/PlanBase.cpp
Comment (triccyx) [reply alica_engine/src/engine/PlanBase.cpp:343]: Yes `Effectively the same as calling reset()`

File: alica_engine/src/engine/RuntimePlanFactory.cpp
Comment (ntfshard) [reply alica_engine/src/engine/RuntimePlanFactory.cpp:44]: Ah, now I see. Yep, I guess I could put all synchronization stuff in a nested scope to allow concurrent method invocation.

File: alica_engine/src/engine/RuntimePlanFactory.cpp
Comment (ntfshard) [reply alica_engine/src/engine/RuntimePlanFactory.cpp]: I believe it could be very deep question. The main difference with predicate is that a predicate is a function, small but function and we also have a call for `wait`, so technically we have 2 calls. I believe it's not a big deal in general for modern CPUs.
For case with while it's just testing one bool w/o calls if it's already in true (like expected in this case)

File: alica_engine/src/engine/RuntimePlanFactory.cpp
Comment (xffox) [inline alica_engine/src/engine/RuntimePlanFactory.cpp:43]: Shouldn't `_creator` usage also be under the lock still?

Comment (ntfshard) [reply alica_engine/src/engine/RuntimePlanFactory.cpp:43]: Initially it was called as-is. And seems inside we have only creation routines.
Strongly speaking I keep old behaviour here (except the entering lock).

UPD: It could be really cool if we can get comment from code owner about it

Comment (xffox) [reply alica_engine/src/engine/RuntimePlanFactory.cpp:43]: Just from the memory correctness perspective i think it should be under the lock if i don't miss anything.

**Review by xffox — APPROVED** [review]

File: alica_engine/src/engine/RuntimePlanFactory.cpp
Comment (ntfshard) [reply alica_engine/src/engine/RuntimePlanFactory.cpp:43]: > Just from the memory correctness perspective i think it should be under the lock if i don't miss anything.

Maybe, but it should be analyzed to make a final decision. It's like if we are only reading something which already was propagated it should be ok.

File: alica_engine/src/engine/RuntimePlanFactory.cpp
Comment (Dale-Koenig) [reply alica_engine/src/engine/RuntimePlanFactory.cpp]: Would also prefer using predicate, if only to make it cleaner.

**Review by Dale-Koenig — APPROVED** [review]

**Pull request merged**
